### PR TITLE
feat(integrations): pre-file dedup gate for ticket outbox

### PR DESCRIPTION
**Pull request by alex-budanov**
Async intelligence dedup runs after the outbox fires, so duplicate bug reports produce duplicate Jira/Linear tickets today. Adds an opt-in per-org grace window (`intelligence_pre_file_dedup_grace_ms`, default 0) that defers the outbox worker until `bug_reports.duplicate_of` has had a chance to settle, then suppresses the external API call if the bug was flagged as a duplicate. Jira/Linear clients still lack `addComment`/`closeIssue`, so retroactive cleanup of already-filed duplicates is a follow-up PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Per-organization pre-filing deduplication grace window to defer ticket filing and re-check duplicates.
  * Outbox entries support a terminal "skipped" state and record a skip reason.

* **Behavior**
  * Worker defers processing during active grace windows, re-checks duplicates, skips entries for missing/duplicate reports, and tracks skipped counts in stats.

* **Tests**
  * Added/updated tests covering grace deferral, skipping, duplicate handling, and related edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/apex-bridge/bugspotter/pull/138)

### CODE REVIEW

**Comment by alex-budanov** [comment]
Addressed in 74d2c0a:
- Removed redundant `bugReports.findById` — preloaded entity now threaded into `createExternalTicket`.
- `scheduled_at = graceUntil ?? new Date()` so the poller waits natively; `deferUntil` stays as a safety net for org-settings changes / retries.
- `getStats()` now reports the `skipped` bucket.

Skipping the `application.` schema-qualification suggestion — every existing query in this repo (`ticket-creation-outbox`, `bug-report`, etc.) is unqualified and relies on `search_path`. Fixing only my new queries would be inconsistent; a sweep is its own PR.

File: packages/backend/src/services/integrations/auto-ticket-service.ts [hunk 1]
```diff
@@ -168,6 +176,7 @@ export class AutoTicketService {
         },
         scheduled_at: new Date(), // Process immediately
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/services/integrations/auto-ticket-service.ts]: Done in 74d2c0a. Kept the worker-side defer as a safety net for org-settings changes / retries.

File: packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts [hunk 2]
```diff
@@ -89,6 +114,23 @@ export class TicketCreationOutboxProcessor {
         scheduledAt: entry.scheduled_at,
       });
 
+      // Step 2.5: Pre-file dedup gate. If the async intelligence pipeline has flagged
+      // this bug as a duplicate of another (duplicate_of != null), don't create a
+      // redundant external ticket — mark the outbox row 'skipped' instead. This is a
+      // pure suppression: bug_reports.duplicate_of is set elsewhere (dedup-service),
+      // and Jira/Linear clients don't expose addComment/closeIssue methods yet, so
+      // we can only avoid the new ticket, not cross-link to the canonical one.
+      const bugReport = await this.db.bugReports.findById(entry.bug_report_id);
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts:94]: Fixed in 74d2c0a — createExternalTicket now accepts the preloaded bugReport.

**Comment by alex-budanov** [comment]
@coderabbitai: getStats now includes skipped (74d2c0a). Skipping the application.schema qualification — every existing query in this repo is unqualified; a sweep belongs in its own PR.

File: packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts [hunk 3]
```diff
@@ -89,9 +115,27 @@ export class TicketCreationOutboxProcessor {
...
 
-      // Step 3: Create ticket on external platform
+      // Step 2.5: Pre-file dedup gate. If the async intelligence pipeline has flagged
+      // this bug as a duplicate of another (duplicate_of != null), don't create a
+      // redundant external ticket — mark the outbox row 'skipped' instead. This is a
+      // pure suppression: bug_reports.duplicate_of is set elsewhere (dedup-service),
+      // and Jira/Linear clients don't expose addComment/closeIssue methods yet, so
+      // we can only avoid the new ticket, not cross-link to the canonical one.
+      const bugReport = await this.db.bugReports.findById(entry.bug_report_id);
+      if (bugReport?.duplicate_of) {
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts]: Good catch — fixed in 3239052: throw immediately at the gate on null, matching createExternalTicket contract.

File: packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts [hunk 4]
```diff
@@ -140,10 +184,14 @@ export class TicketCreationOutboxProcessor {
...
-   * Create ticket on external platform using plugin registry
+   * Create ticket on external platform using plugin registry.
+   *
+   * Optionally accepts a pre-fetched bug report to avoid an extra DB round-trip
+   * when the caller has already loaded it (e.g. the dedup gate above).
    */
   private async createExternalTicket(
-    entry: TicketCreationOutboxEntry
+    entry: TicketCreationOutboxEntry,
+    preloadedBugReport?: BugReport | null
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts]: Done in 3239052 — narrowed to BugReport | undefined now that the null case is rejected upstream.

File: packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts [hunk 5]
```diff
@@ -58,6 +59,31 @@ export class TicketCreationOutboxProcessor {
     });
 
     try {
+      // Step 0: If the entry has a pre-file dedup grace window that hasn't elapsed yet,
+      // push scheduled_at forward and exit cleanly. The poller will re-enqueue the
+      // entry after the grace window closes. Done BEFORE markProcessing so we don't
+      // burn a retry attempt on an entry that simply isn't due yet.
+      const pre = await this.db.ticketOutbox.findById(outboxEntryId);
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts]: Done in 1ab7aee — markProcessing now does UPDATE ... RETURNING *; defensive findById is gone.

File: packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts [hunk 6]
```diff
@@ -58,6 +59,31 @@ export class TicketCreationOutboxProcessor {
...
+      if (!pre) {
+        logger.error('Outbox entry not found in database', { outboxEntryId, jobId: job.id });
+        throw new Error(`Outbox entry not found: ${outboxEntryId}`);
+      }
+      if (pre.dedup_grace_until && new Date() < pre.dedup_grace_until && pre.status === 'pending') {
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts]: Fixed in 1ab7aee — grace check now covers both pending and failed, deferUntil widened to match.

File: packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts [hunk 7]
```diff
@@ -89,9 +107,33 @@ export class TicketCreationOutboxProcessor {
...
+      // pure suppression: bug_reports.duplicate_of is set elsewhere (dedup-service),
+      // and Jira/Linear clients don't expose addComment/closeIssue methods yet, so
+      // we can only avoid the new ticket, not cross-link to the canonical one.
+      const bugReport = await this.db.bugReports.findById(entry.bug_report_id);
+      if (!bugReport) {
+        // Match createExternalTicket's contract — surface the missing bug as a hard
+        // failure so the outbox retry path takes over (vs. silently falling through
+        // and triggering a second findById inside createExternalTicket).
+        throw new Error(`Bug report not found: ${entry.bug_report_id}`);
+      }
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts:106]: Good call — fixed in 4852ebc. Missing bug is now terminal via markSkipped(bug_report_not_found); no retry burn.

File: packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts [hunk 8]
```diff
@@ -58,26 +59,43 @@ export class TicketCreationOutboxProcessor {
...
+          logger.debug('Outbox entry deferred for pre-file dedup grace', {
+            outboxEntryId,
+            graceUntil: pre.dedup_grace_until,
+            status: pre.status,
+          });
+          return;
+        }
+        // Race: another worker claimed the entry between our findById and
+        // deferUntil — fall through to the normal claim path so its state wins.
+      }
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/queue/workers/outbox/ticket-creation-outbox.worker.ts]: Confirmed and fixed in 4852ebc. Rather than route around the BullMQ semantics with a moveToDelayed leak through the message-broker abstraction, removed the worker-side defer entirely — grace is enforced at row creation via scheduled_at, the duplicate_of gate is the real suppression.
